pkg/endpoint/bpf.go

```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 		e.PolicyDebug(logrus.Fields{"dumpedDiffs": diffs}, "syncPolicyMapWithDump")
 	}
 
+	if err == nil {
+		e.policyMapLastFullReconciliation = time.Now()
+	}
+
 	return err
 }
 
-func (e *Endpoint) startSyncPolicyMapController() {
-	// Skip the controller if the endpoint has no policy map
-	if !e.HasBPFPolicyMap() {
-		return
+func (e *Endpoint) syncPolicyMapControllerName() string {
+	return fmt.Sprintf("sync-policymap-%d", e.ID)
+}
+
+// updateSyncPolicyMapController creates or triggers a controller to  reconcile
+// pending policy map changes. Every PolicyMapFullReconciliationInterval the
+// controller also does full reconciliation by dumping the policy map.
+func (e *Endpoint) updateSyncPolicyMapController() channels.DoneChan {
+	// Skip the controller if the endpoint has no policy map or is
+	// being disconnected.
+	if !e.HasBPFPolicyMap() || e.IsDisconnecting() {
+		return channels.ClosedDoneChan
 	}
 
-	ctrlName := fmt.Sprintf("sync-policymap-%d", e.ID)
-	e.controllers.CreateController(ctrlName,
+	return e.controllers.UpdateController(
+		e.syncPolicyMapControllerName(),
 		controller.ControllerParams{
 			Group:          syncPolicymapControllerGroup,
 			HealthReporter: e.GetReporter("policymap-sync"),
 			DoFunc: func(ctx context.Context) error {
-				// Failure to lock is not an error, it means
-				// that the endpoint was disconnected and we
-				// should exit gracefully.
-				if err := e.lockAlive(); err != nil {
+				e.unconditionalLock()
+				defer e.unlock()
+				if e.IsDisconnecting() {
 					return controller.NewExitReason("Endpoint disappeared")
 				}
-				defer e.unlock()
+
 				if e.desiredPolicy != e.realizedPolicy {
 					// Currently in the middle of a regeneration; do not execute
 					// at this time.
 					return nil
 				}
-				return e.syncPolicyMapWithDump()
+
+				err := e.syncPolicyMapWithDump(ctx)
```

Comment:
Another divergence -- before, `ep.ApplyPolicyMapChanges()` does **not** dump, it only applies deltas. That makes it much more efficient. The background controller does an expensive map dump to make sure there hasn't been divergence. This change now adds this dump to the FQDN-response critical path, which is very latency sensitive.

We really need to not do map dumps during FQDN responses. This probably means we need two controllers :-(.

Reply:
Right, even if we have this check to limit the rate of full reconciliation with dump:

```
if time.Since(e.policyMapLastFullReconciliation) < option.Config.PolicyMapFullReconciliationInterval {
	return nil
}
```

It would not be nice to hit that in the FQDN path. Let's use a separate controller, then.

Reply:
Created a dedicated controller for incremental updates.